### ISSUE TITLE
support exception-free build

### ISSUE THREAD

**Issue by alyssawilk** [envoyproxy/envoy#27412]
We've slowly whittled down exceptions from the Envoy data plane but even with the work we've done on https://github.com/envoyproxy/envoy/issues/14320 there are still some exceptions in the data plane, and it's hard for new users of Envoy to know where to use exceptions and where to use other error handling.

I'd like to move towards Envoy being able to build without exceptions entirely, primarily for https://github.com/envoyproxy/envoy-mobile/issues/176 but also due to continued concerns in-house with performance and safety of exceptions.

I have a proof-of-concept "no exceptions" build of Envoy Mobile, which currently replaces "throw" with "throw or panic" in about 60 files (down from ~120, and envoy mobile panics on config parse fail in any case), and straight removal of catch macros.  I'd like to 
1) disallow new core code files from throwing exceptions (which we can add to iff necessary)
2) replace existing catch calls with CATCH_EXCEPTION_MAIN_THREAD{}  which can compile out in no-exceptions build
3) slowly replace the throwEnvoyExceptionOrPanic call sites with statusOr returns boiling up to base server creation
4) as each plug-in gets a "throw free" creation mode, disallow new extensions of that type from throwing on creation

we'll probably have to keep the TRY/CATCH macros in Envoy code perpetually, but it'd be really fantastic to get at least core code building without it, and maybe some of the folks who are eager to do clean up work can work on moving older extensions to statusor mode.

**Comment by yanavlasov** [comment]
This will solve a lot of internal problems, trying to mix exception enabled Envoy code with exception disabled internal code.  However it is going to a massive refactor.

**Comment by htuch** [comment]
I agree on the positives. The main challenge in core code is that it is a very common pattern to use exceptions during config ingestion for validation. There is no reason this code could not eventually move to an exception free model, but it would require a significant effort to switch to a pattern like `statusOr` across the board. Hopefully someone has the cycles for this work.

Will we also ask extensions to move away from exceptions once core is exception free? One concern is that currently exceptions are an "invisible" (i.e. not observable at compile time) part of the extension API interface.

**Comment by alyssawilk** [comment]
they're invisible by API but if you try to build fnoexceptions and you've got a lingering throw around the compiler will 100% catch it for you.  As I've discovered a dozen times over the last month merging my no_exception branch =P

Should be pretty easy to switch the E-M build to no exceptions, then hopefully (but who knows)  some day the core build.  I don't think it's worth ever switching contrib, or asking folks downstream of Envoy to rework.  it can be a permanent bazel option to compile out or in the TRY/CATCH macros

**Comment by adisuissa** [comment]
It is a non-trivial amount of work, but if there are benefits, then it may be a worthy cause.

I suggest adding files (maybe even directories) to [this list](https://github.com/envoyproxy/envoy/blob/a6703d5cf446d209a7d9b450ab2b6c23d376b6ab/tools/code_format/config.yaml#L87) whenever that file is clean.

**Comment by alyssawilk** [comment]
cc @paul-r-gall re 285574162 we can only dream....
